### ISSUE TITLE
repeatable motion with cursor-moving

### ISSUE THREAD

**Issue by mattn** [vim/vim#5542]
Currently, vim does not allow dot-repeat with motion include moving-cursor.

```
imap ( ()<left>
```
When typing `(<cr>`, vim insert:

```
(
)
```
But `.` does not repleat inserting this text since `<left>` stop recording. What about to add way to keep recording with moving cursor? ex: `inoremap <cursor>`

**Comment by chrisbra** [comment]
not sure I understand, but does `:h i_CTRL-G_U` help you?

**Comment by mattn** [comment]
Could you please try this https://github.com/mattn/vim-lexiv/

with small fix.

```diff
diff --git a/autoload/lexiv.vim b/autoload/lexiv.vim
index ded7dc7..7320d52 100644
--- a/autoload/lexiv.vim
+++ b/autoload/lexiv.vim
@@ -38,7 +38,7 @@ function! lexiv#paren_open(lhs) abort
       return a:lhs	
     endif
   endif
-  return l:lhs . l:rhs . "\<left>"
+  return l:lhs . l:rhs . "\<c-g>U\<left>"
 endfunction
 
 function! lexiv#paren_expand() abort
```

`(` expand `()` but not dot repeatable.

**Comment by mattn** [comment]
`:imap <expr> ( "()\<c-g>U\<left>"` seems working well but if the rhs is returned from function, it does not work.

**Comment by chrisbra** [comment]
from my limited testing, it seemed to work.

**Comment by mattn** [comment]
Simple test should work. But `("` (`("")`) does not.

**Comment by mattn** [comment]
Sorry, this seems my confusing.

**Comment by mattn** [comment]
@chrisbra I could implement vim-lexiv as dot-repeatable but when inserting `<CR>`, it does not work.

```
(<cr><cr><cr>
```
![terminal7](https://user-images.githubusercontent.com/10111/73225362-c0537a80-41af-11ea-82fb-09a03afed942.gif)

**Comment by chrisbra** [comment]
Can you please show a simple reproduction mapping?

**Comment by mattn** [comment]
```
inoremap <expr> { "{}\<c-g>U\<left>"
inoremap <expr> <cr> "\<cr>\<c-g>U\<c-o>O"
```

And types
```
o<tab>{<cr><esc>gg.
```
I expect two blocks `{...}` are there but `.` insert just single empty line.

**Comment by chrisbra** [comment]
Well, the whole point of `<C-G>U` was to only allow join the undo information, if the cursor moves **within** a line. If you want to move to other positions, it will intentionally not work. The reason being, that else we might have problems with undo/redo storing the wrong undo information (which could lead to undo-corruption and leave the buffer with wrong text). This was discussed in length on the vim-dev list (and possibly also here), when I initially created that patch (sorry, don't find the link currently)

So no, there is no way around it and it is intentional.

**Comment by mattn** [comment]
Yes, I know. Thanks. So I'm suggesting to add an identifier to enable this feature.

**Comment by mattn** [comment]
@chrisbra Moving to right or left are allowed. Why up or down are not?

**Comment by chrisbra** [comment]
because this will break undo

**Comment by brammool** [comment]
I can understand that it would be very nice if this works, so you can map {  to appending an empty line and the }, then move the cursor up a line, user types some test, and that whole sequence can be repeated like when the } was typed last.  But I'm not sure how to make that work in general.
